legate/core/context.py

```diff
@@ -66,6 +67,36 @@ def find_last_user_frame(libname: str) -> str:
     return f"{frame.f_code.co_filename}:{frame.f_lineno}"
 
 
+class LibraryAnnotations:
+    def __init__(self) -> None:
+        self._entries: dict[str, str] = {}
+        self._provenance: Union[str, None] = None
+
+    @property
+    def provenance(self) -> Optional[str]:
+        return self._provenance
+
+    def set_provenance(self, provenance: str) -> None:
+        self._provenance = provenance
+
+    def reset_provenance(self) -> None:
+        self._provenance = None
+
+    def update(self, **kwargs: Any) -> None:
+        self._entries.update(**kwargs)
+
+    def remove(self, key: str) -> None:
+        del self._entries[key]
+
+    def __repr__(self) -> str:
+        pairs = (f"{key},{value}" for key, value in self._entries.items())
+        return "|".join(
+            pairs
+            if self._provenance is None
+            else chain(pairs, (f"Provenance,{self._provenance}",))
+        )
```

Comment:
maybe a little simpler, and avoids the `itertools` import:

```suggestion
        if self._provenance is not None:
            pairs += (f"Provenance,{self._provenance}",)
        return "|".join(pairs)
```

Reply:
ok I didn't know you could add a tuple to a generator, but wouldn't that materialize the generator into some sequence?

Reply:
looks like I can't even do it
```
TypeError: unsupported operand type(s) for +=: 'generator' and 'tuple'
```

Reply:
I mean, I'm fine with materializing the generator into a list, I guess I was trying to be too clever...

Reply:
oh right, you'd have to call `tuple` on it first, either way seems fine